File: oio/conscience/checker/http.py
# Copyright (C) 2015-2020 OpenIO SAS, as part of OpenIO SDS
# Copyright (C) 2022-2023 OVH SAS
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3.0 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library.

from oio.common import exceptions as exc
from oio.common.constants import REQID_HEADER
from oio.common.http_urllib3 import urllibexc
from oio.conscience.checker.base import BaseChecker


class HttpChecker(BaseChecker):
    checker_type = "http"

    def _configure(self):
        for k in ("uri",):
            if k not in self.checker_conf:
                raise exc.ConfigurationException(
                    f'Missing field "{k}" in configuration'
                )

        self.path = self.checker_conf["uri"].lstrip("/")
        self.name = f"{self.name}|{self.path}"
        self.url = "%s:%s%s%s" % (
            self.host,
            self.port,
            "" if self.path.startswith("/") else "/",
            self.path,
        )

    def _check(self, reqid=None):
        resp = None
        try:
            # We have clues that the connection will be reused quickly to get
            # stats, thus we do not explicitly require its closure.
            hdrs = {REQID_HEADER: reqid}
            resp = self.agent.pool_manager.request("GET", self.url, headers=hdrs)
            if resp.status == 200:
                self.last_check_success = True
            else:
                raise Exception(f"({resp.status}) {resp.data.decode('utf-8')}")
        except Exception as err:
            # Avoid spamming the logs
            if self.last_check_success:
                self.logger.warn(
                    "ERROR performing %s check (%s reqid=%s): %s",
                    self.checker_type,
                    self.url,
                    reqid,
                    err,
                )
            self.last_check_success = False
        finally:
            if resp:
                try:
                    # Probably useless since we do not disable preload_content
                    resp.drain_conn()
                    resp.close()
                except urllibexc.HTTPError:
                    pass
            if not self.last_check_success:
                self.logger.warn("%s check failed (reqid=%s)", self.name, reqid)
            return self.last_check_success

File: rawx/http.go
package main

// OpenIO SDS Go rawx
// Copyright (C) 2015-2020 OpenIO SAS
// Copyright (C) 2023-2024 OVH SAS
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Affero General Public
// License as published by the Free Software Foundation; either
// version 3.0 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with this program. If not, see <http://www.gnu.org/licenses/>.

import (
	"fmt"
	"net"
	"net/http"
	"os"
	"strconv"
	"sync"

	"openio-sds/rawx/logger"
)

func Run(wg *sync.WaitGroup, srv *httpServer, tls_cert_file string, tls_key_file string) error {
	var protocol string
	var fd int64 = -1
	var addr string
	var err error
	var listener net.Listener

	if tls_cert_file != "" && tls_key_file != "" {
		protocol = "HTTPS"
	} else {
		protocol = "HTTP"
	}

	fdEnvVar := os.Getenv(fmt.Sprintf("__OIO_RAWX_FORK_%s_FD", protocol))
	if fdEnvVar != "" {
		fd, err = strconv.ParseInt(fdEnvVar, 10, 64)
		if err != nil {
			logger.LogWarning("[%s] Error while parsing __OIO_RAWX_FORK_%s_FD env variable: %v", protocol, protocol, err)
			fd = -1
		}
	}
	if fd > -1 {
		addr = os.Getenv(fmt.Sprintf("__OIO_RAWX_FORK_%s_ADDR", protocol))
		if addr == "" || addr != srv.server.Addr {
			logger.LogWarning("[%s] graceful restart asked but Addr changed, use a new socket (old=%s new=%s)", protocol, addr, srv.server.Addr)
			file := os.NewFile(uintptr(fd), "")
			file.Close()
			fd = -1
		}
	}

	if fd > -1 {
		logger.LogInfo("[%s] About to use fd %d to listen on %s", protocol, fd, srv.server.Addr)
		file := os.NewFile(uintptr(fd), "")
		listener, err = net.FileListener(file)
		if err != nil {
			fd = -1
			return fmt.Errorf("[%s] Unable to listen (%s) using existing FD %d: %v", protocol, srv.server.Addr, fd, err)
		}
	}
	if fd == -1 {
		listener, err = net.Listen("tcp", srv.server.Addr)
		if err != nil {
			return err
		}
	}
	srv.socket = listener.(*net.TCPListener)

	wg.Add(1)

	go func() {
		defer wg.Done()

		var err error

		logger.LogInfo("[%s] About to serve on %s", protocol, srv.server.Addr)
		if tls_cert_file != "" && tls_key_file != "" {
			err = srv.server.ServeTLS(listener, tls_cert_file, tls_key_file)
		} else {
			err = srv.server.Serve(listener)
		}

		if err != http.ErrServerClosed {
			logger.LogWarning("[%s] Unable to start server: %v", protocol, err.Error())
		}

		logger.LogInfo("[%s] server stopped on %s", protocol, srv.server.Addr)
	}()

	return nil
}
